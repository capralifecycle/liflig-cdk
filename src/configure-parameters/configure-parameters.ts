import * as iam from "@aws-cdk/aws-iam"
import * as secretsmanager from "@aws-cdk/aws-secretsmanager"
import * as ssm from "@aws-cdk/aws-ssm"
import * as cdk from "@aws-cdk/core"
import * as crypto from "crypto"

/**
 * A plain text application parameter that will be stored
 * in AWS Parameter Store. Only used for non-sensitive values,
 * typically those commited directly in the IaC code.
 */
export interface PlainTextParameter {
  key: string
  value: string
}

/**
 * An AWS Secret that should hold a JSON object.
 *
 * This will provide a secret that liflig-properties can read. The final
 * parameter seen by the application will be the key given here and the
 * keys from the secret JSON object appended.
 */
export interface SecretParameter {
  key: string
  secret: secretsmanager.ISecret
}

/**
 * An AWS Secret that should hold a JSON object.
 *
 * This will provide a secret that liflig-properties can read. The final
 * parameter seen by the application will be the key given here and the
 * keys from the secret JSON object appended.
 */
export interface SecretByNameParameter {
  key: string
  secretName: string
}

export type Parameter =
  | PlainTextParameter
  | SecretParameter
  | SecretByNameParameter

interface Props {
  /**
   * Prefix used for parameter names.
   * Should start with '/' and end without '/'.
   */
  ssmPrefix: string
  parameters: Parameter[]
}

export class ConfigureParameters {
  public parameters: ssm.IParameter[]
  public ssmPrefix: string
  private secrets: secretsmanager.ISecret[]

  private configParameters: PlainTextParameter[]
  private secretParameters: SecretParameter[]

  constructor(scope: cdk.Construct, props: Props) {
    if (!props.ssmPrefix.startsWith("/") || props.ssmPrefix.endsWith("/")) {
      throw new Error("ssmPrefix should start with '/' and end without '/'")
    }

    this.ssmPrefix = props.ssmPrefix

    this.configParameters = []
    this.secretParameters = []

    for (const parameter of props.parameters) {
      if ("value" in parameter) {
        this.configParameters.push(parameter)
      } else if ("secret" in parameter) {
        this.secretParameters.push(parameter)
      } else if ("secretName" in parameter) {
        this.secretParameters.push({
          key: parameter.key,
          secret: secretsmanager.Secret.fromSecretNameV2(
            scope,
            `SecretRef${parameter.key}`,
            parameter.secretName,
          ),
        })
      }
    }

    const result: ssm.IParameter[] = []

    this.configParameters.forEach((it) => {
      const param = new ssm.StringParameter(scope, `Config${it.key}`, {
        type: ssm.ParameterType.STRING,
        stringValue: it.value,
        parameterName: `${this.ssmPrefix}/config/${it.key}`,
      })
      result.push(param)
    })

    this.secretParameters.forEach((it) => {
      const param = new ssm.StringParameter(scope, `Secret${it.key}`, {
        type: ssm.ParameterType.STRING,
        stringValue: it.secret.secretArn,
        parameterName: `${this.ssmPrefix}/secrets/${it.key}`,
      })
      result.push(param)
    })

    this.secrets = this.secretParameters.map((it) => it.secret)

    this.parameters = result
  }

  grantRead(grantable: iam.IGrantable): void {
    grantable.grantPrincipal.addToPrincipalPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ["ssm:GetParametersByPath"],
        resources: [
          `arn:aws:ssm:${cdk.Stack.of(grantable).region}:${
            cdk.Stack.of(grantable).account
          }:parameter${this.ssmPrefix}/*`,
        ],
      }),
    )

    for (const secret of this.secrets) {
      secret.grantRead(grantable)
    }
  }

  /**
   * Produce a checksum-value that can be used as a kind of
   * nonce to trigger redeployment on parameter changes.
   *
   * Note: If the parameter references a token no change will be visible,
   * and a manual redeployment might be needed.
   */
  get hashValue(): string {
    const hash = crypto.createHash("sha256")

    if (this.configParameters) {
      this.configParameters.forEach((it) => {
        hash.update("|")
        hash.update(it.key)
        hash.update("|")

        if (!cdk.Token.isUnresolved(it.value)) {
          hash.update(it.value)
        }
      })
    }

    if (this.secretParameters) {
      this.secretParameters.forEach((it) => {
        hash.update("|")
        hash.update(it.key)
        hash.update("|")

        if (!cdk.Token.isUnresolved(it.secret.secretArn)) {
          hash.update(it.secret.secretArn)
        }
      })
    }

    return hash.digest("hex")
  }
}
