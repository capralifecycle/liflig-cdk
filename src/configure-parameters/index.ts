export {
  ConfigureParameters,
  ConfigureParametersProps,
  PlainTextParameter,
  SecretParameter,
  SecretByNameParameter,
  Parameter,
} from "./configure-parameters"
