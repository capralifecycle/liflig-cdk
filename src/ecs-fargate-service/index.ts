export {
  EcsFargateService,
  EcsFargateServiceProps,
} from "./ecs-fargate-service"
