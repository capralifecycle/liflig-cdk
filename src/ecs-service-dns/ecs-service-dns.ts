import * as elb from "@aws-cdk/aws-elasticloadbalancingv2"
import * as route53 from "@aws-cdk/aws-route53"
import * as route53Targets from "@aws-cdk/aws-route53-targets"
import * as cdk from "@aws-cdk/core"

export interface EcsServiceDnsProps {
  httpsListener: elb.IApplicationListener
  loadBalancer: elb.IApplicationLoadBalancer
  domainName: string
  listenerPriority: number
  targetGroup: elb.IApplicationTargetGroup
  hostedZone?: route53.IHostedZone
}

export class EcsServiceDns extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props: EcsServiceDnsProps) {
    super(scope, id)

    new elb.ApplicationListenerRule(this, "ListenerRule", {
      listener: props.httpsListener,
      priority: props.listenerPriority,
      hostHeader: props.domainName,
      targetGroups: [props.targetGroup],
    })

    if (props.hostedZone != null) {
      new route53.ARecord(this, "ARecord", {
        zone: props.hostedZone,
        recordName: `${props.domainName}.`,
        target: route53.RecordTarget.fromAlias(
          new route53Targets.LoadBalancerTarget(props.loadBalancer),
        ),
      })
    }
  }
}
