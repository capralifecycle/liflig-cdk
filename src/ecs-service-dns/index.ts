export { EcsServiceDns } from "./ecs-service-dns"
