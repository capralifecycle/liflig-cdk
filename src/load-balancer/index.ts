export { LoadBalancer } from "./load-balancer"
export type { LoadBalancerProps } from "./load-balancer"
