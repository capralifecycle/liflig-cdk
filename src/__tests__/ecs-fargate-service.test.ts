import "@aws-cdk/assert/jest"
import { App, Stack } from "@aws-cdk/core"
import * as ec2 from "@aws-cdk/aws-ec2"
import * as ecs from "@aws-cdk/aws-ecs"
import * as sm from "@aws-cdk/aws-secretsmanager"
import * as ecr from "@aws-cdk/aws-ecr"
import * as cm from "@aws-cdk/aws-certificatemanager"
import "jest-cdk-snapshot"
import { HostedZoneWithParam } from "../hosted-zone-with-param"
import { LoadBalancer } from "../load-balancer"
import { EcsFargateService } from "../ecs-fargate-service"
import { EcsServiceDns } from "../ecs-service-dns"
import { Parameter } from "../configure-parameters/configure-parameters"

test("creates fargate service with parameters and DNS", () => {
  const app = new App()
  const stack = new Stack(app, "Stack")

  const vpc = new ec2.Vpc(stack, "Vpc", {
    subnetConfiguration: [
      {
        cidrMask: 19,
        name: "Public",
        subnetType: ec2.SubnetType.PUBLIC,
      },
    ],
  })

  const hostedZoneWithParam = new HostedZoneWithParam(stack, "HostedZone", {
    zoneName: "example.com",
  })

  const hostedZone = hostedZoneWithParam.getHostedZone(stack, "HostedZone")

  const certificate = new cm.Certificate(stack, "Certificate", {
    domainName: `*.example.com`,
    subjectAlternativeNames: ["example.com"],
    validation: cm.CertificateValidation.fromDns(hostedZone),
  })

  const loadBalancer = new LoadBalancer(stack, "LoadBalancer", {
    certificates: [certificate],
    vpc: vpc,
  })

  const ecsCluster = new ecs.Cluster(stack, "Cluster", {
    vpc,
  })

  const ecrRepository = new ecr.Repository(stack, "Repository", {
    repositoryName: "example-repository",
  })

  const parameters: Parameter[] = [
    {
      key: "ExamplePlainTextParameter",
      value: "ExamplePlainTextParameter",
    },
    {
      key: "ExampleSecretParameter",
      secret: sm.Secret.fromSecretNameV2(
        stack,
        "ExampleSecretParameter",
        "ExampleSecretParameter",
      ),
    },
    {
      key: "ExampleSecretReferenceParameter",
      secretName: "SecretReferenceParameter",
    },
  ]

  const service = new EcsFargateService(stack, "Service", {
    serviceName: "example-service",
    vpc: vpc,
    cluster: ecsCluster,
    desiredCount: 2,
    parameters,
    ecrRepository: ecrRepository,
    ecrTag: "exampleEcrTag",
  })

  new EcsServiceDns(stack, "Dns", {
    domainName: `example.com`,
    hostedZone: hostedZone,
    httpsListener: loadBalancer.httpsListener,
    listenerPriority: 10,
    loadBalancer: loadBalancer.loadBalancer,
    targetGroup: service.targetGroup,
  })

  expect(stack).toMatchCdkSnapshot()
})
