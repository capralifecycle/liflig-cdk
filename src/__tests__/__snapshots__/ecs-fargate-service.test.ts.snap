// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates fargate service with parameters and DNS 1`] = `
Object {
  "Resources": Object {
    "Certificate4E7ABB08": Object {
      "Properties": Object {
        "DomainName": "*.example.com",
        "DomainValidationOptions": Array [
          Object {
            "DomainName": "example.com",
            "HostedZoneId": Object {
              "Ref": "HostedZoneDB99F866",
            },
          },
        ],
        "SubjectAlternativeNames": Array [
          "example.com",
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "ClusterEB0386A7": Object {
      "Type": "AWS::ECS::Cluster",
    },
    "DnsARecordFED6ED67": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::Join": Array [
              "",
              Array [
                "dualstack.",
                Object {
                  "Fn::GetAtt": Array [
                    "LoadBalancer8E47CE52",
                    "DNSName",
                  ],
                },
              ],
            ],
          },
          "HostedZoneId": Object {
            "Fn::GetAtt": Array [
              "LoadBalancer8E47CE52",
              "CanonicalHostedZoneID",
            ],
          },
        },
        "HostedZoneId": Object {
          "Ref": "HostedZoneDB99F866",
        },
        "Name": "example.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "DnsListenerRule18B182CC": Object {
      "Properties": Object {
        "Actions": Array [
          Object {
            "TargetGroupArn": Object {
              "Ref": "ServiceTargetGroupE74A3394",
            },
            "Type": "forward",
          },
        ],
        "Conditions": Array [
          Object {
            "Field": "host-header",
            "Values": Array [
              "example.com",
            ],
          },
        ],
        "ListenerArn": Object {
          "Ref": "LoadBalancerHttpsListener68D81C94",
        },
        "Priority": 10,
      },
      "Type": "AWS::ElasticLoadBalancingV2::ListenerRule",
    },
    "HostedZoneDB99F866": Object {
      "Properties": Object {
        "Name": "example.com.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "LoadBalancer8E47CE52": Object {
      "DependsOn": Array [
        "VpcPublicSubnet1DefaultRoute3DA9E72A",
        "VpcPublicSubnet2DefaultRoute97F91067",
      ],
      "Properties": Object {
        "LoadBalancerAttributes": Array [
          Object {
            "Key": "deletion_protection.enabled",
            "Value": "false",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "LoadBalancerSecurityGroupC89CA14B",
              "GroupId",
            ],
          },
        ],
        "Subnets": Array [
          Object {
            "Ref": "VpcPublicSubnet1Subnet5C2D37C4",
          },
          Object {
            "Ref": "VpcPublicSubnet2Subnet691E08A3",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "LoadBalancerDefaultTargetGroup162BFC2A": Object {
      "Properties": Object {
        "Port": 80,
        "Protocol": "HTTP",
        "TargetType": "instance",
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "LoadBalancerHttpListener9DD37EAA": Object {
      "Properties": Object {
        "DefaultActions": Array [
          Object {
            "RedirectConfig": Object {
              "Port": "443",
              "Protocol": "HTTPS",
              "StatusCode": "HTTP_301",
            },
            "Type": "redirect",
          },
        ],
        "LoadBalancerArn": Object {
          "Ref": "LoadBalancer8E47CE52",
        },
        "Port": 80,
        "Protocol": "HTTP",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "LoadBalancerHttpsListener68D81C94": Object {
      "Properties": Object {
        "Certificates": Array [
          Object {
            "CertificateArn": Object {
              "Ref": "Certificate4E7ABB08",
            },
          },
        ],
        "DefaultActions": Array [
          Object {
            "TargetGroupArn": Object {
              "Ref": "LoadBalancerDefaultTargetGroup162BFC2A",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": Object {
          "Ref": "LoadBalancer8E47CE52",
        },
        "Port": 443,
        "Protocol": "HTTPS",
        "SslPolicy": "ELBSecurityPolicy-TLS-1-2-2017-01",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "LoadBalancerSecurityGroupC89CA14B": Object {
      "Properties": Object {
        "GroupDescription": "Automatically created Security Group for ELB StackLoadBalancerF0AF61AA",
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 443",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "LoadBalancerSecurityGrouptoStackServiceSecurityGroupC1D3B07B8080D942545E": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "ServiceSecurityGroupC96ED6A7",
            "GroupId",
          ],
        },
        "FromPort": 8080,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "LoadBalancerSecurityGroupC89CA14B",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 8080,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "Repository22E53BBD": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RepositoryName": "example-repository",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Retain",
    },
    "Service9571FDD8": Object {
      "DependsOn": Array [
        "DnsListenerRule18B182CC",
        "ServiceConfigExamplePlainTextParameterC948638B",
        "ServiceSecretExampleSecretParameter2147C87C",
        "ServiceSecretExampleSecretReferenceParameter9AB84512",
      ],
      "Properties": Object {
        "Cluster": Object {
          "Ref": "ClusterEB0386A7",
        },
        "DeploymentConfiguration": Object {
          "MaximumPercent": 200,
          "MinimumHealthyPercent": 100,
        },
        "DesiredCount": 2,
        "EnableECSManagedTags": false,
        "HealthCheckGracePeriodSeconds": 60,
        "LaunchType": "FARGATE",
        "LoadBalancers": Array [
          Object {
            "ContainerName": "Container",
            "ContainerPort": 8080,
            "TargetGroupArn": Object {
              "Ref": "ServiceTargetGroupE74A3394",
            },
          },
        ],
        "NetworkConfiguration": Object {
          "AwsvpcConfiguration": Object {
            "AssignPublicIp": "ENABLED",
            "SecurityGroups": Array [
              Object {
                "Fn::GetAtt": Array [
                  "ServiceSecurityGroupC96ED6A7",
                  "GroupId",
                ],
              },
            ],
            "Subnets": Array [
              Object {
                "Ref": "VpcPublicSubnet1Subnet5C2D37C4",
              },
              Object {
                "Ref": "VpcPublicSubnet2Subnet691E08A3",
              },
            ],
          },
        },
        "PlatformVersion": "1.4.0",
        "ServiceName": "example-service",
        "TaskDefinition": Object {
          "Ref": "ServiceTaskDefinition55FA0F15",
        },
      },
      "Type": "AWS::ECS::Service",
    },
    "ServiceConfigExamplePlainTextParameterC948638B": Object {
      "Properties": Object {
        "Name": "/liflig-cdk/Stack/c8cabff989df74f9b2755f4ae923cc6492f944970b/parameters/config/ExamplePlainTextParameter",
        "Type": "String",
        "Value": "ExamplePlainTextParameter",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ServiceLogGroupB910EE76": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RetentionInDays": 14,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "ServiceSecretExampleSecretParameter2147C87C": Object {
      "Properties": Object {
        "Name": "/liflig-cdk/Stack/c8cabff989df74f9b2755f4ae923cc6492f944970b/parameters/secrets/ExampleSecretParameter",
        "Type": "String",
        "Value": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":secretsmanager:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":secret:ExampleSecretParameter",
            ],
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ServiceSecretExampleSecretReferenceParameter9AB84512": Object {
      "Properties": Object {
        "Name": "/liflig-cdk/Stack/c8cabff989df74f9b2755f4ae923cc6492f944970b/parameters/secrets/ExampleSecretReferenceParameter",
        "Type": "String",
        "Value": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":secretsmanager:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":secret:SecretReferenceParameter",
            ],
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ServiceSecurityGroupC96ED6A7": Object {
      "Properties": Object {
        "GroupDescription": "Stack/Service/SecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "ServiceSecurityGroupfromStackLoadBalancerSecurityGroupA6CA8EF5808028DAA90F": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "FromPort": 8080,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "ServiceSecurityGroupC96ED6A7",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "LoadBalancerSecurityGroupC89CA14B",
            "GroupId",
          ],
        },
        "ToPort": 8080,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "ServiceTargetGroupE74A3394": Object {
      "Properties": Object {
        "HealthCheckIntervalSeconds": 10,
        "HealthCheckPath": "/health",
        "HealthyThresholdCount": 2,
        "Port": 8080,
        "Protocol": "HTTP",
        "TargetGroupAttributes": Array [
          Object {
            "Key": "deregistration_delay.timeout_seconds",
            "Value": "15",
          },
        ],
        "TargetType": "ip",
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "ServiceTaskDefinition55FA0F15": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Environment": Array [
              Object {
                "Name": "SSM_PREFIX",
                "Value": "/liflig-cdk/Stack/c8cabff989df74f9b2755f4ae923cc6492f944970b/parameters",
              },
              Object {
                "Name": "PARAMS_HASH",
                "Value": "3a807a41646dc1f972e5cf6a70bb7938",
              },
            ],
            "Essential": true,
            "Image": Object {
              "Fn::Join": Array [
                "",
                Array [
                  Object {
                    "Fn::Select": Array [
                      4,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "Repository22E53BBD",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                  ".dkr.ecr.",
                  Object {
                    "Fn::Select": Array [
                      3,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "Repository22E53BBD",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                  ".",
                  Object {
                    "Ref": "AWS::URLSuffix",
                  },
                  "/",
                  Object {
                    "Ref": "Repository22E53BBD",
                  },
                  ":exampleEcrTag",
                ],
              ],
            },
            "LogConfiguration": Object {
              "LogDriver": "awslogs",
              "Options": Object {
                "awslogs-datetime-format": "%Y-%m-%dT%H:%M:%S",
                "awslogs-group": Object {
                  "Ref": "ServiceLogGroupB910EE76",
                },
                "awslogs-region": Object {
                  "Ref": "AWS::Region",
                },
                "awslogs-stream-prefix": "ecs",
              },
            },
            "Name": "Container",
            "PortMappings": Array [
              Object {
                "ContainerPort": 8080,
                "HostPort": 8080,
                "Protocol": "tcp",
              },
            ],
          },
        ],
        "Cpu": "256",
        "ExecutionRoleArn": Object {
          "Fn::GetAtt": Array [
            "ServiceTaskDefinitionExecutionRole5D3DB197",
            "Arn",
          ],
        },
        "Family": "StackServiceTaskDefinition81B5CCD5",
        "Memory": "512",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": Array [
          "FARGATE",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "ServiceTaskDefinitionTaskRole3BD32B0F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "ServiceTaskDefinitionExecutionRole5D3DB197": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ServiceTaskDefinitionExecutionRoleDefaultPolicyE16C69DE": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "Repository22E53BBD",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "ecr:GetAuthorizationToken",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "ServiceLogGroupB910EE76",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ServiceTaskDefinitionExecutionRoleDefaultPolicyE16C69DE",
        "Roles": Array [
          Object {
            "Ref": "ServiceTaskDefinitionExecutionRole5D3DB197",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ServiceTaskDefinitionTaskRole3BD32B0F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ServiceTaskDefinitionTaskRoleDefaultPolicyED991BA3": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/liflig-cdk/Stack/c8cabff989df74f9b2755f4ae923cc6492f944970b/parameters/*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":secretsmanager:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":secret:ExampleSecretParameter-??????",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":secretsmanager:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":secret:SecretReferenceParameter-??????",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ServiceTaskDefinitionTaskRoleDefaultPolicyED991BA3",
        "Roles": Array [
          Object {
            "Ref": "ServiceTaskDefinitionTaskRole3BD32B0F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Vpc8378EB38": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VpcIGWD7BA715C": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "VpcPublicSubnet1DefaultRoute3DA9E72A": Object {
      "DependsOn": Array [
        "VpcVPCGWBF912B6E",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet1RouteTable6C95E38E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPublicSubnet1RouteTable6C95E38E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Vpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcPublicSubnet1RouteTableAssociation97140677": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet1RouteTable6C95E38E",
        },
        "SubnetId": Object {
          "Ref": "VpcPublicSubnet1Subnet5C2D37C4",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcPublicSubnet1Subnet5C2D37C4": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "Stack/Vpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcPublicSubnet2DefaultRoute97F91067": Object {
      "DependsOn": Array [
        "VpcVPCGWBF912B6E",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet2RouteTable94F7E489",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPublicSubnet2RouteTable94F7E489": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Vpc/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcPublicSubnet2RouteTableAssociationDD5762D8": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet2RouteTable94F7E489",
        },
        "SubnetId": Object {
          "Ref": "VpcPublicSubnet2Subnet691E08A3",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcPublicSubnet2Subnet691E08A3": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.32.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "Stack/Vpc/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcVPCGWBF912B6E": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
}
`;
